File: 0733-flood-fill/0733-flood-fill.py
from typing import List

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        rows, cols = len(image), len(image[0])
        startColor = image[sr][sc]
        
        if startColor == newColor:  # no need to change
            return image

        def dfs(r, c):
            if r < 0 or c < 0 or r >= rows or c >= cols:
                return
            if image[r][c] != startColor:
                return
            image[r][c] = newColor
            dfs(r+1, c)
            dfs(r-1, c)
            dfs(r, c+1)
            dfs(r, c-1)

        dfs(sr, sc)
        return image

        

File: 0733-flood-fill/0733-flood-fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} color
 * @return {number[][]}
 */
var floodFill = function(image, sr, sc, color) {
    let cols = image[0].length
    let rows = image.length
    let originalColor = image[sr][sc]
    if(originalColor === color){
        return image
    }
    function dfs(r,c){
        if(r<0 || c<0 || r >= rows || c >= cols || image[r][c] !== originalColor){
            return 
        }
        image[r][c] =color
        dfs(r,c-1)
        dfs(r-1,c)
        dfs(r,c+1)
        dfs(r+1,c)
    }
    dfs(sr,sc)
    return image
};